File: golang/103.binary-tree-zigzag-level-order-traversal.go
/*
 * @lc app=leetcode id=103 lang=golang
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 */

package golang

// @lc code=start
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
// func zigzagLevelOrder(root *TreeNode) [][]int {
// 	left := false
// 	tree := zigzagTree([][]*TreeNode{{root}}, left)
// 	return genArray(tree)
// }

// func zigzagTree(treeSlice [][]*TreeNode, left bool) [][]*TreeNode {
// 	// Placeholder for lint, not an actual code
// 	if treeSlice == nil {
// 		return [][]int{}
// 	}
// 	result := make([][]int, 0)
// 	cur := make([]int, 0)
// 	nextTreeSlice := make([]*TreeNode, 0)
// 	if left {
// 		for _, tree := range treeSlice {
// 			if tree == nil {
// 				continue
// 			}
// 			cur = append(cur, tree.Val)
// 			nextTreeSlice = append(nextTreeSlice, tree.Left, tree.Right)
// 		}
// 	}
// 	result = append(result, cur)
// 	return append(result, zigzagTree(nextTreeSlice, !left)...)
// }

// @lc code=end

File: typescript/src/103.binary-tree-zigzag-level-order-traversal.ts
/*
 * @lc app=leetcode id=103 lang=typescript
 *
 * [103] Binary Tree Zigzag Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     val: number
 *     left: TreeNode | null
 *     right: TreeNode | null
 *     constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
 *         this.val = (val===undefined ? 0 : val)
 *         this.left = (left===undefined ? null : left)
 *         this.right = (right===undefined ? null : right)
 *     }
 * }
 */

function helpZigzagLevelOrder({
  root,
  depth,
  result,
}: {
  root: TreeNode | null
  depth: number
  result: number[][]
}): void {
  if (root === null) {
    return
  }
  if (result.length === depth) {
    result.push([])
  }
  if (depth % 2 === 0) {
    result[depth].push(root.val)
  } else {
    result[depth].unshift(root.val)
  }
  helpZigzagLevelOrder({
    root: root.left,
    depth: depth + 1,
    result,
  })
  helpZigzagLevelOrder({
    root: root.right,
    depth: depth + 1,
    result,
  })
}

function zigzagLevelOrder(root: TreeNode | null): number[][] {
  const result: number[][] = []
  helpZigzagLevelOrder({
    root,
    depth: 0,
    result,
  })
  return result
}
// @lc code=end
